tensorflow/contrib/framework/python/ops/accumulate_n_v2.py

```diff
@@ -0,0 +1,113 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Ops that will eventually be folded into tensorflow/python/ops/math_ops.py
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.contrib.framework.python.ops import gen_accumulate_n_v2_ops
+from tensorflow.contrib.util import loader
+
+from tensorflow.python.eager import context
+from tensorflow.python.framework import ops
+from tensorflow.python.platform import resource_loader
+from tensorflow.python.framework import tensor_shape
+
+
+def accumulate_n_v2(inputs, shape=None, tensor_dtype=None, name=None):
+  """Returns the element-wise sum of a list of tensors.
+
+  Optionally, pass `shape` and `tensor_dtype` for shape and type checking,
+  otherwise, these are inferred.
+
+  `tf.accumulate_n_v2` performs the same operation as `tf.add_n`, but does not
+  wait for all of its inputs to be ready before beginning to sum. This can
+  save memory if inputs are ready at different times, since minimum temporary
+  storage is proportional to the output size rather than the inputs size.
+
+  Unlike the original `accumulate_n`, `accumulate_n_v2` is differentiable.
+
+  For example:
+
+  ```python
+  a = tf.constant([[1, 2], [3, 4]])
+  b = tf.constant([[5, 0], [0, 6]])
+  tf.accumulate_n_v2([a, b, a])  # [[7, 4], [6, 14]]
+
+  # Explicitly pass shape and type
+  tf.accumulate_n_v2([a, b, a], shape=[2, 2], tensor_dtype=tf.int32)  
+                                                                   # [[7,  4],
+                                                                   #  [6, 14]]
+  ```
+
+  Args:
+    inputs: A list of `Tensor` objects, each with same shape and type.
+    shape: Shape of elements of `inputs`.
+    tensor_dtype: The type of `inputs`.
+    name: A name for the operation (optional).
+
+  Returns:
+    A `Tensor` of same shape and type as the elements of `inputs`.
+
+  Raises:
+    ValueError: If `inputs` don't all have same shape and dtype or the shape
+    cannot be inferred.
+  """
+  loader.load_op_library(
+      resource_loader.get_path_to_datafile("_accumulate_n_v2_op.so"))
+  _INPUTS_ERR_MSG = ValueError("inputs must be a list of at least one Tensor"
+                               "with the same dtype and shape")
+  if context.in_eager_mode():
+    # TODO(apassos) remove this once the lifetime of eager variables gets
+    # addressed.
+    raise ValueError("accumulate_n not supported in eager mode")
```

Comment:
Better yet, just use add_n here.

Reply:
Good idea. I'll put a fallback to `add_n` for eager mode at the bottom of this function, after the argument validation code. I'll also add a test case modeled after `OpsTest.testExecuteIntAttr()` in `python/eager/ops_test.py`.